types/src/v0/impls/stake_table.rs

```diff
@@ -26,89 +30,254 @@ use hotshot_types::{
+                };
+
+                // The contract does *not* enforce that each schnorr key is only used once.
+                if schnorr_keys.contains(&state_ver_key) {
+                    tracing::warn!("schnorr key {} already used", state_ver_key.to_string());
+                };
+
+                bls_keys.insert(stake_table_key);
+                schnorr_keys.insert(state_ver_key.clone());
+
+                match validators.entry(account) {
+                    indexmap::map::Entry::Occupied(_occupied_entry) => {
+                        bail!("validator {:#x} already registered", *account)
+                    },
+                    indexmap::map::Entry::Vacant(vacant_entry) => vacant_entry.insert(Validator {
+                        account,
+                        stake_table_key,
+                        state_ver_key,
+                        stake: U256::from(0_u64),
+                        commission,
+                        delegators: HashMap::default(),
+                    }),
+                };
+            },
+            StakeTableEvent::Deregister(exit) => {
+                validators
+                    .shift_remove(&exit.validator)
+                    .with_context(|| format!("validator {:#x} not found", exit.validator))?;
+            },
+            StakeTableEvent::Delegate(delegated) => {
+                let Delegated {
+                    delegator,
+                    validator,
+                    amount,
+                } = delegated;
+                let validator_entry = validators
+                    .get_mut(&validator)
+                    .with_context(|| format!("validator {validator:#x} not found"))?;
+
+                if amount.is_zero() {
+                    tracing::warn!("delegator {delegator:?} has 0 stake");
+                    continue;
+                }
+                // Increase stake
+                validator_entry.stake += amount;
+                // Add delegator to the set
+                validator_entry.delegators.insert(delegator, amount);
+            },
+            StakeTableEvent::Undelegate(undelegated) => {
+                let Undelegated {
+                    delegator,
+                    validator,
+                    amount,
+                } = undelegated;
+                let validator_entry = validators
+                    .get_mut(&validator)
+                    .with_context(|| format!("validator {validator:#x} not found"))?;
+
+                validator_entry.stake = validator_entry
+                    .stake
+                    .checked_sub(amount)
+                    .with_context(|| "stake is less than undelegated amount")?;
+
+                let delegator_stake = validator_entry
+                    .delegators
+                    .get_mut(&delegator)
+                    .with_context(|| format!("delegator {delegator:#x} not found"))?;
+                *delegator_stake = delegator_stake.checked_sub(amount).unwrap();
```

Comment:
This should not be unwrap but return an error like above `delegated amount less than undelegated amount`

Reply:
https://github.com/EspressoSystems/espresso-network/pull/2772/commits/fc8d14dc3780c6d5295723e136e92aba8d1ec2bb